books/RayTracingTheNextWeek.html

```diff
@@ -2057,105 +2057,147 @@
+            // defined, looks only in that directory for the image file. If RTW_IMAGES is not
+            // defined, this constructor will search for the specified image file first from the
+            // current directory, then in the images/ subdirectory, then the _parent's_ images/
+            // subdirectory, and then _that_ parent, on so on, for six levels up. If the image was
+            // not loaded successfully, width() and height() will return 0.
+
+            const auto imagedir = getenv("RTW_IMAGES");
+            if (imagedir) {
+                // If RTW_IMAGES is defined, then only use that value to locate the image file.
+                if (load(std::string(imagedir) + "/" + image_filename)) return;
+            } else {
+                // Hunt for the image file in some likely locations.
+                auto filename = std::string(image_filename);
+                if (load(filename)) return;
+                if (load("images/" + filename)) return;
+                if (load("../images/" + filename)) return;
+                if (load("../../images/" + filename)) return;
+                if (load("../../../images/" + filename)) return;
+                if (load("../../../../images/" + filename)) return;
+                if (load("../../../../../images/" + filename)) return;
+                if (load("../../../../../../images/" + filename)) return;
             }
 
-            bytes_per_scanline = bytes_per_pixel * width;
-        }
-
-        ~image_texture() {
-            delete data;
+            std::cerr << "ERROR: Could not load image file '" << image_filename << "'.\n";
+            std::cerr << "STBI: " << stbi_failure_reason() << '\n';
         }
 
-        virtual color value(double u, double v, const vec3& p) const override {
-            // If we have no texture data, then return solid cyan as a debugging aid.
-            if (data == nullptr)
-                return color(0,1,1);
+        ~rtw_image() { STBI_FREE(data); }
 
-            // Clamp input texture coordinates to [0,1] x [1,0]
-            u = interval(0,1).clamp(u);
-            v = 1.0 - interval(0,1).clamp(v);  // Flip V to image coordinates
+        bool load(const std::string filename) {
+            // Loads image data from the given file name. Returns true if the load succeeded.
+            auto components_per_pixel = bytes_per_pixel; // Components per pixel
+            data = stbi_load(filename.c_str(),
+                &image_width, &image_height, &components_per_pixel, bytes_per_pixel);
+            bytes_per_scanline = image_width * bytes_per_pixel;
+            return data != nullptr;
+        }
 
-            auto i = static_cast<int>(u * width);
-            auto j = static_cast<int>(v * height);
+        int width()  const { return (data == nullptr) ? 0 : image_width; }
+        int height() const { return (data == nullptr) ? 0 : image_height; }
```

Comment:
maybe return -1?

Reply:
what does stbi return if it fails to load?

Reply:
STBI returns `nullptr` on load failure.

Also, I want to return zero, so the single magenta indicator pixel still works. Basically, if the load fails, but you don't really care, everything will still work, with a magenta color. I don't want to leave this in a zombie state.